namastack-outbox-tracing/src/main/kotlin/io/namastack/outbox/OutboxHandlerTracingAspect.kt

```
@@ -40,35 +43,36 @@ class OutboxHandlerTracingAspect(
 
-        return runWithSpan(span) { joinPoint.proceed() }
+        return runWithSpan(spanBuilder, record) { joinPoint.proceed() }
     }
 
     /**
-     * Executes block within span context with proper error handling and cleanup.
+     * Executes a block within a span context with proper error handling and cleanup.
      *
-     * Makes span current so nested operations inherit the trace context.
-     * Records exceptions and sets error status before rethrowing.
-     * Ensures span is ended even if exception occurs.
+     * Makes the span current so nested operations inherit the trace context.
+     * If the outbox record has a failure exception, it is recorded in the span
+     * before the span is ended. Ensures the span is always properly closed
+     * even if an exception occurs during execution.
      *
-     * @param span Consumer span to execute within
-     * @param block Code to execute within span context
+     * @param T The return type of the block
+     * @param spanBuilder Span builder to create and execute within
+     * @param record Outbox record being processed, used to check for failure exceptions
+     * @param block Lambda to execute within the span context
      * @return Result of block execution
-     * @throws Exception Rethrows any exception after recording it in span
      */
     private inline fun <T> runWithSpan(
-        span: Span,
+        spanBuilder: Span.Builder,
+        record: OutboxRecord<*>,
         block: () -> T,
-    ): T =
+    ): T {
+        val span = spanBuilder.start()
         try {
-            span.makeCurrent().use { block() }
-        } catch (ex: Exception) {
-            span.recordException(ex)
-            span.setStatus(ERROR)
-
-            throw ex
+            return tracer.withSpan(span).use { block() }
         } finally {
+            record.failureException?.let { error -> span.error(error) }
             span.end()
         }
+    }
```

This is personal taste, but I like the old version much more, because it is very explicit:

1. Run Code in Span
2. Catch all exceptions
3. Set Span Status and Exception
4. Rethrow Exception
5. Finally end the span

---

You're right in terms of throwing expcetions in the processing chain. These are being catched and not rethrowed. Thanks for pointing this out!

Nethertheless, I think that the following implementation is correct.

```kotlin
private inline fun <T> runWithSpan(
        span: Span,
        block: () -> T,
    ): T =
        try {
            tracer.withSpan(span).use { block() }
        } catch (ex: Exception) {
            span.error(ex)
            throw ex
        } finally {
            span.end()
        }
```

Let me explain why:

1. Framework vs. Business Separation: The span represents framework processing (try → retry → fallback → mark failed), not business outcomes. The framework succeeds even when business logic fails - this is working as designed.
2. Exception-Driven Error Model: The try-catch only marks errors when exceptions propagate out, indicating infrastructure failures (database crashes, configuration errors) rather than expected business failures handled through retries/fallbacks.
3. Clear and Safe: Activate span context, execute code, catch framework failures, always end span. This matches OpenTelemetry best practices and ensures spans are properly closed.
4. Correct Observability: Business failures are observable through outbox.record.failure_count tags, database status, and logs. Marking the framework span as "error" for expected business failures would create false alerts.
5. Messaging Semantics: In async messaging, consumer spans represent message delivery and framework processing, not business success. A properly handled failure (stored as FAILED) is successful framework operation.
6. User Handler Observability: When users annotate handlers with `@Observed`, those create child spans that automatically show errors on exceptions. This creates proper hierarchy: parent outbox process span shows framework success, child handler span shows business logic failure - clearly distinguishing infrastructure issues from business errors handled through retries/fallbacks.